test/unit/modules/blocks/utils.js

```diff
@@ -13,268 +13,581 @@
+		b_height: 3,
+		t_id: '6950874693022090568',
+		t_type: 0,
+	},
+];
+
 describe('blocks/utils', () => {
-	describe('Utils', () => {
-		describe('library', () => {
-			it('should assign logger');
+	var dbStub;
+	var loggerStub;
+	var blockMock;
+	var transactionMock;
+	var accountMock;
+	var blocksUtilsModule;
+	var modulesStub;
+	var __private;
+	var library;
+	var modules;
+
+	beforeEach(() => {
+		dbStub = {
+			blocks: {
+				getIdSequence: sinonSandbox.stub().resolves(),
+				getHeightByLastId: sinonSandbox.stub().resolves(['1']),
+				loadLastBlock: sinonSandbox.stub().resolves(viewRow_full_blocks_list),
+				loadBlocksData: sinonSandbox.stub(),
+				aggregateBlocksReward: sinonSandbox.stub().resolves(),
+			},
+		};
+
+		dbStub.blocks.loadBlocksData
+			.withArgs(sinonSandbox.match({ id: '13068833527549895884' }))
+			.resolves(viewRow_full_blocks_list)
+			.withArgs(sinonSandbox.match({ id: '1' }))
+			.resolves([]);
+
+		loggerStub = {
+			trace: sinonSandbox.spy(),
+			info: sinonSandbox.spy(),
+			error: sinonSandbox.spy(),
+		};
+
+		blockMock = {
+			dbRead: function(input) {
+				return { id: input.b_id, height: input.b_height };
+			},
+		};
+
+		transactionMock = {
+			dbRead: function(input) {
+				return { id: input.t_id, type: input.t_type };
+			},
+		};
+
+		accountMock = {
+			get: sinonSandbox.stub(),
+		};
+
+		accountMock.get
+			.withArgs(sinonSandbox.match({ address: 'ERRL' }))
+			.callsArgWith(1, 'Address error stub', null)
+			.withArgs(sinonSandbox.match({ address: '0L' }))
+			.callsArgWith(1, null, undefined)
+			.withArgs(sinonSandbox.match({ address: '1L' }))
+			.callsArgWith(1, null, { publicKey: '123abc' });
+
+		blocksUtilsModule = new BlocksUtils(
+			loggerStub,
+			accountMock,
+			blockMock,
+			transactionMock,
+			dbStub,
+			modulesLoader.scope.dbSequence,
+			modulesLoader.scope.genesisblock
+		);
 
-			it('should assign db');
+		modulesStub = {
+			blocks: {
+				lastBlock: {
+					get: sinonSandbox
+						.stub()
+						.returns({ id: '9314232245035524467', height: 1 }),
+					set: sinonSandbox
+						.stub()
+						.returns({ id: '9314232245035524467', height: 1 }),
+				},
+				utils: {
+					readDbRows: blocksUtilsModule.readDbRows,
+				},
+			},
+		};
+
+		blocksUtilsModule.onBind(modulesStub);
+
+		__private = BlocksUtils.__get__('__private');
+		__private.loaded = false;
+
+		library = BlocksUtils.__get__('library');
+		modules = BlocksUtils.__get__('modules');
+	});
 
-			it('should assign dbSequence');
+	afterEach(() => {
+		sinonSandbox.reset();
+	});
 
-			it('should assign genesisblock');
+	describe('constructor', () => {
+		describe('library', () => {
+			it('should assign logger', () => {
+				expect(library.logger).to.eql(loggerStub);
+			});
 
 			describe('should assign logic', () => {
-				it('should assign block');
+				it('should assign account', () => {
+					expect(library.logic.account).to.eql(accountMock);
+				});
+
+				it('should assign block', () => {
```

Comment:
again, those tests are not reliable

we should test like that:
- create instance
    - should assign logger
    - should assign account
    - should assign block
    - etc.

and not like that:
- create instance
    - should assign logger
- create instance
    - should assign account
- create instance
    - should assign block
- create instance
    - etc.

because we have no guaranties that it will ever assign all those properties on single instance together.

I finish the review here for now.

Reply:
I disagree, when it comes to adding further examples (further test coverage in the future), it is more reliable and also more maintainable. Each test example works by itself, atomically.

Reply:
I did a mistake by stopping review here, rest of tests looks good.

What I meant that creating instance is single operation here (there is only one path in constructor), so if we want to recreate instance in `beforeEach`, then we should test constructor like that:
- beforeEach
  - create instance
- describe constructor
  - should create proper/valid instance, log message and return self
          // should assign transaction logic
    - expect(library.logic.transaction).to.eql(transactionMock);
      // should assign dbSequence
    - expect(library.dbSequence).to.eql(modulesLoader.scope.dbSequence);
      // should call library.logger.trace with "Blocks->Utils: Submodule initialized."
    - expect(loggerStub.trace.args[0][0]).to.equal('Blocks->Utils: Submodule initialized.');
      // should return self
    - expect(blocksUtilsModule).to.be.an('object');
    - expect(blocksUtilsModule.readDbRows).to.be.a('function');
    - etc.

Because there is only one path - testing that path should be done in single instance. Like it's done in other parts of that tests.

Another approach:
- describe constructor
  - before
    - recreateInstance()
  - should assign transaction logic
    - expect(library.logic.transaction).to.eql(transactionMock);
  - should assign dbSequence
    - expect(library.dbSequence).to.eql(modulesLoader.scope.dbSequence);
  - should call library.logger.trace with "Blocks->Utils: Submodule initialized."
    - expect(loggerStub.trace.args[0][0]).to.equal('Blocks->Utils: Submodule initialized.');
  - should return self
    - expect(blocksUtilsModule).to.be.an('object');
    - expect(blocksUtilsModule.readDbRows).to.be.a('function');
  - etc.

In most cases approach like:
- describe first path
  - create instance (beforeEach)
    - test random element of first path
  - create instance (beforeEach)
    - test some other element of first path

should be valid, as behavior should be repliceable if tests are written good (first path is always the same path), but imo it's not guaranteed and we should avoid that approach as not reliable.

Reply:
NOTE: This comment is just fo the reference get some point of view. 

We are making an instance of of module by:
```
blocksUtilsModule = new BlocksUtils(
  loggerStub,
  accountMock,
  blockMock,
  transactionMock,
  dbStub,
  modulesLoader.scope.dbSequence,
  modulesLoader.scope.genesisblock
);
```
After creating an instance with that specific number of arguments, then in unit tests we should test the possible control flows; 

1. If there is dynamic assignment of variables then have multiple `it` cases
2. If there is a conditional constructor then have multiple `it` cases for each condition

Just testing straight forward sequence of lines without dynamicty and any condition in multiple `it` blocks does not really makes sense to me and also does not improve the test coverage. So what I prefer; 

- describe('constructor')
   - it('should properly assign construct params')
       - expect param 1
       - expect param2
       - expect param3

So the rule of thumb should be:

### ONE **it** should test ONE **flow/path** in a function/constructor call 

And if one flow/path contains multiple expectations then make them part of that it.

Reply:
Done according to your stated principle @nazarhussain